### PR TITLE
CSPL-320: Smartstore Config

### PR DESCRIPTION
**Pull request by sgontla**
- First cut changes just support Smartstore config.
- More changes to come, to support the overal feature

### CODE REVIEW

File: deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml [hunk 1]
```diff
@@ -2309,13 +2364,37 @@ spec:
...
+                  type: array
+              type: object
           type: object
       type: object
   version: v1alpha3
   versions:
   - name: v1alpha3
     served: true
     storage: true
-  - name: v1alpha2
```
Comment (mikedickey) [inline deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml:2319]: Why remove v1alpha2? I think this prevents someone from upgrading v2 to v3.

File: pkg/apis/enterprise/v1alpha3/common_types.go [hunk 2]
```diff
@@ -67,3 +67,25 @@ type CommonSplunkSpec struct {
...
 	IndexerClusterRef corev1.ObjectReference `json:"indexerClusterRef"`
 }
+
+// SmartStoreSpec defines Splunk indexes and remote storage volume configuration
+type SmartStoreSpec struct {
+	// List to accommodate multiple volumes.
+	// For now, a hard check is added such that only one volume is supported to accommodate one common secrete object
+	S3VolList []S3VolumeSpec `json:"volumes,ommitempty"`
+
+	IndexList []IndexSpec `json:"indexes,ommitempty"`
```
Comment (mikedickey) [inline pkg/apis/enterprise/v1alpha3/common_types.go:77]: please comment each public field. I'm also surprised/concerned that "make lint" isn't failing this?

File: pkg/apis/enterprise/v1alpha3/common_types.go [hunk 3]
```diff
@@ -67,3 +67,25 @@ type CommonSplunkSpec struct {
...
+}
+
+// S3VolumeSpec defines the remote storage volume configuration
+type S3VolumeSpec struct {
+	RemoteName string `json:"volumeName"`
```
Comment (mikedickey) [inline pkg/apis/enterprise/v1alpha3/common_types.go:82]: field naming is inconsistent across these. I think since it's in the context of a "S3VolumeSpec" that just "Name"/"name" is sufficient.

File: pkg/apis/enterprise/v1alpha3/common_types.go [hunk 4]
```diff
@@ -67,3 +67,25 @@ type CommonSplunkSpec struct {
...
+	S3Endpoint string `json:"endpoint"`
+	// Some optional config like API version etc..?
+}
+
+// IndexSpec defines Splunk indexes configuration
+type IndexSpec struct {
+	Indexname      string `json:"indexName"`
```
Comment (mikedickey) [inline pkg/apis/enterprise/v1alpha3/common_types.go:89]: Same here -- I think just "Name"/"name" would be better

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go [hunk 5]
```diff
@@ -88,6 +94,9 @@ type IndexerClusterStatus struct {
 
 	// status of each indexer cluster peer
 	Peers []IndexerClusterMemberStatus `json:"peers"`
+
+	// Splunk Smart store configuration
+	SmartStore SmartStoreSpec `json:"smartstore,ommitempty"`
```
Comment (mikedickey) [inline pkg/apis/enterprise/v1alpha3/indexercluster_types.go:99]: think you have an extra "m" here: "omitempty"

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go [hunk 6]
```diff
@@ -55,6 +58,9 @@ type IndexerClusterMemberStatus struct {
 
 	// Flag indicating if this peer belongs to the current committed generation and is searchable.
 	Searchable bool `json:"is_searchable"`
+
+	// Splunk Smart store configuration
+	SmartStore SmartStoreSpec `json:"smartstore,ommitempty"`
```
Comment (mikedickey) [inline pkg/apis/enterprise/v1alpha3/indexercluster_types.go:63]: why do we need this in member status?

File: pkg/splunk/enterprise/configuration.go [hunk 7]
```diff
@@ -606,3 +609,63 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
 		podTemplateSpec.Spec.Containers[idx].Env = env
 	}
 }
+
+// ListSmartStoreVolumes lists smartstore volumes
```
Comment (mikedickey) [inline pkg/splunk/enterprise/configuration.go:613]: I think "Log*" may be a better name for these vs more generic "List*", to help clarify the expected outcome

File: pkg/splunk/enterprise/configuration.go [hunk 8]
```diff
@@ -606,3 +609,63 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
...
+func ListSmartStoreIndexes(indexList []enterprisev1.IndexSpec) {
+	scopedLog := logC.WithName("ListSmartStoreIndexes")
+	for _, index := range indexList {
+		scopedLog.Info("Index: ", "IndexName: ", index.Indexname, "RemoteLocation: ", index.RemoteLocation)
+	}
+}
+
+// isSmartstoreEnabled checks and returns true if smartstore is configured
+func isSmartstoreConfigured(smartstore *enterprisev1.SmartStoreSpec) bool {
+	return !(smartstore.IndexList == nil && smartstore.S3VolList == nil)
```
Comment (mikedickey) [inline pkg/splunk/enterprise/configuration.go:632]: what if smartstore == nil?

File: pkg/splunk/enterprise/configuration.go [hunk 9]
```diff
@@ -606,3 +609,63 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
...
+}
+
+// ValidateSplunkSmartstoreSpec checks and validates the smartstore config
+func ValidateSplunkSmartstoreSpec(smartstore *enterprisev1.SmartStoreSpec) error {
+	if !isSmartstoreConfigured(smartstore) {
+		return nil
+	}
+	numVolumes := len(smartstore.S3VolList)
+	if numVolumes <= 0 {
+		return fmt.Errorf("Smartstore Volume configueration is missing")
```
Comment (mikedickey) [inline pkg/splunk/enterprise/configuration.go:642]: typo configuration

File: deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml (hunk 1 above)
Comment (sgontla) [reply deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml:2319]: Thanks Mike. Yesterday, we were discussing this.  Actually, I didn't do this change, and it was auto-generated when i generated the CRDs with the smartstore config. I will revert it.

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go (hunk 5 above)
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/indexercluster_types.go:99]: Thanks for this nice catch Mike. I am thinking if this is the reason for the CircleCI failure. (Wondering, if it is not supposed to be caught by the compiler)

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go (hunk 6 above)
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/indexercluster_types.go:63]: I thought we can use this to check if there is a change in the smartstore config from last reconcile iteration to the now, so that, we can create a new config map. 
Please correct me, if my understanding is wrong in using this way, and if there needs to be a better alternative way.

File: deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml (hunk 1 above)
Comment (mikedickey) [reply deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml:2319]: Understood. Thought I added something to build/package.sh so that it would add this back into each of the CRD yamls, but looks like I did not. Would you be able to look into that? I think it'll just take a few extra "yq" commands, and then we don't have to worry about it getting clobbered again (we'll need to update it for each version bump, though).

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go (hunk 6 above)
Comment (mikedickey) [reply pkg/apis/enterprise/v1alpha3/indexercluster_types.go:63]: Ahh ok, that makes sense. Generally, it's better to compare to the state of the output. For example, most things we compare to the current spec of the corresponding StatefulSet objects that are produced. In this case, I imagine the "output" would be the ConfigMap that is the app for indexes.conf.

Comparing to indexes.conf seems overly cumbersome. An alternative may be to store this as an extra field in the ConfigMap. The approach of storing it in status is probably fine, though.

File: deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml (hunk 1 above)
Comment (sgontla) [reply deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml:2319]: Sure Mike, i will take a look at it

File: pkg/apis/enterprise/v1alpha3/common_types.go (hunk 2 above)
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/common_types.go:77]: Sorry, i missed to reply earlier. No, make lint is not failing.

File: pkg/splunk/enterprise/configuration.go (hunk 8 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration.go:632]: > what if smartstore == nil?

I wanted to verify through debugger once to be sure before I reply. I just checked, and since SmartStore is a struct object, eve if the smartstore spec is not configured, it will not become nil, but the values it holds can be nil.

Comment (sgontla) [reply pkg/splunk/enterprise/configuration.go:632]: @mikedickey  Verified through the debugger, and smartstore never becomes nil even if we don't specify in the CR. SmartStore is a struct object(not dynamically allocated) in the  StandaloneSpec and IndexerClusterSpec data structures.

File: pkg/apis/enterprise/v1alpha3/indexercluster_types.go (hunk 6 above)
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/indexercluster_types.go:63]: Will evaluate and take a direction as part of CSPL-348, which is in progress for createing the configmap.

File: deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml [hunk 10]
```diff
@@ -1002,6 +1002,38 @@ spec:
...
+                        type: string
+                      name:
+                        description: Splunk data index name
+                        type: string
+                    type: object
+                  type: array
+                volumes:
+                  description: List to accommodate multiple volumes. For now, a hard
+                    check is added such that only one volume is supported to accommodate
+                    one common secrete object
```
Comment (smohan-splunk) [inline deploy/crds/enterprise.splunk.com_indexerclusters_crd.yaml:1024]: Nitpick - typo with secret

File: pkg/splunk/enterprise/configuration.go [hunk 11]
```diff
@@ -639,3 +642,69 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
...
+	}
+
+	return !(smartstore.IndexList == nil && smartstore.VolList == nil)
+}
+
+// ValidateSplunkSmartstoreSpec checks and validates the smartstore config
+func ValidateSplunkSmartstoreSpec(smartstore *enterprisev1.SmartStoreSpec) error {
+	// Smartstore is an optional config (at least) for now
+	if !isSmartstoreConfigured(smartstore) {
+		return nil
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration.go:676]: Its better to log an error msg saying something like  - "Smartstore is not configured" or something along those lines.

File: pkg/splunk/enterprise/indexercluster.go [hunk 12]
```diff
@@ -322,12 +329,23 @@ func getClusterMasterStatefulSet(cr *enterprisev1.IndexerCluster) (*appsv1.State
...
+
 func validateIndexerClusterSpec(cr *enterprisev1.IndexerCluster) error {
 	// IndexerCluster must support 0 replicas for multisite / multipart clusters
 	// Multisite / multipart clusters: can't reference a cluster master located in another namespace because of Service and Secret limitations
 	if len(cr.Spec.IndexerClusterRef.Namespace) > 0 && cr.Spec.IndexerClusterRef.Namespace != cr.GetNamespace() {
 		return fmt.Errorf("Multisite cluster does not support cluster master to be located in a different namespace")
 	}
+
+	if cr.Spec.IndexerClusterRef.Name == "" {
+		err := ValidateSplunkSmartstoreSpec(&cr.Spec.SmartStore)
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/indexercluster.go:341]: if `smartstore==null`, `ValidateSplunkSmartstoreSpec` should return an error and we should return from here, right?

File: pkg/splunk/enterprise/indexercluster.go [hunk 13]
```diff
@@ -322,12 +329,23 @@ func getClusterMasterStatefulSet(cr *enterprisev1.IndexerCluster) (*appsv1.State
...
+		if err != nil {
+			return err
+		}
+	} else if isSmartstoreConfigured(&cr.Spec.SmartStore) {
+		return (fmt.Errorf("Smartstore configuration is not supporeted on this site. Smartstore can be configured only on site with CM"))
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/indexercluster.go:346]: **Comment 1.** 
Nit: s/supporeted/supported
**Comment 2.**  
I don't think so we should mention the word "site" here as it might confuse people with multi-site feature. Maybe use node or pod or resource?

File: pkg/splunk/enterprise/configuration_test.go [hunk 14]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
 		}
 	}
 }
+
+func TestSmartstoreApplyIndexerClusterFailsOnInvalidSmartStoreConfi(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:240]: nit: s/TestSmartstoreApplyIndexerClusterFailsOnInvalidSmartStoreConfi/TestSmartstoreApplyIndexerClusterFailsOnInvalidSmartStoreConfig

File: pkg/splunk/enterprise/configuration_test.go [hunk 15]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+	cr := enterprisev1.IndexerCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "standalone",
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:243]: Do we want to change the name from "standalone" to something else, since this is not exactly testing smartstore on standalone instance, right?

File: pkg/splunk/enterprise/configuration_test.go [hunk 16]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+			Namespace: "test",
+		},
+		Spec: enterprisev1.IndexerClusterSpec{
+			Replicas: 1,
+			SmartStore: enterprisev1.SmartStoreSpec{
+				VolList: []enterprisev1.VolumeSpec{
+					{Name: "msos_s2s3_vol", S3Endpoint: ""},
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:250]: So this test case should fail because `S3Endpoint: ""` ?

File: pkg/splunk/enterprise/configuration_test.go [hunk 17]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+
+	var client splcommon.ControllerClient
+
+	_, err := ApplyIndexerCluster(client, &cr)
+	if err == nil {
+		t.Errorf("ApplyIndexerCluster should fail on invalid smartstore config")
+	}
+}
+
+func TestSmartstoreApplyStandaloneFailsOnInvalidSmartStoreConfi(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:270]: nit: s/TestSmartstoreApplyStandaloneFailsOnInvalidSmartStoreConfi/TestSmartstoreApplyStandaloneFailsOnInvalidSmartStoreConfig

File: pkg/splunk/enterprise/configuration_test.go [hunk 18]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+
+	var client splcommon.ControllerClient
+
+	_, err := ApplyStandalone(client, &cr)
+	if err == nil {
+		t.Errorf("ApplyStandalone should fail on invalid smartstore config")
+	}
+}
+
+func TestSmartStoreConfigDoNotFailsOnSitesWithCM(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:300]: Please do not use "site" if its not related to multi-site env.

File: pkg/splunk/enterprise/configuration_test.go [hunk 19]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+	cr := enterprisev1.IndexerCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "idx_CM_site",
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:303]: ditto, same as above!

File: pkg/splunk/enterprise/configuration_test.go [hunk 20]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+	}
+
+	err := validateIndexerClusterSpec(&cr)
+
+	if err != nil {
+		t.Errorf("Smartstore configuration is Invalid: %v", err)
+	}
+}
+
+func TestSmartStoreConfigFailsOnSitesWithNoCM(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:329]: ditto: don't use "site" or "sites"

File: pkg/splunk/enterprise/configuration_test.go [hunk 21]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+	cr := enterprisev1.IndexerCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "idexer_site_1",
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:332]: ditto:same as above

File: pkg/splunk/enterprise/configuration_test.go [hunk 22]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+				IndexList: []enterprisev1.IndexSpec{
+					{Name: "salesdata1", RemoteLocation: "testbucket-rs-london/$_index_name"},
+					{Name: "salesdata2", RemoteLocation: "testbucket-rs-london/$_index_name"},
+					{Name: "salesdata3", RemoteLocation: "testbucket-rs-london/$_index_name"},
+				},
+			},
+		},
+	}
+
+	cr.Spec.IndexerClusterRef.Name = "testCMSiteName"
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:351]: ditto:same as above

File: pkg/splunk/enterprise/configuration_test.go [hunk 23]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+
+	err := validateIndexerClusterSpec(&cr)
+
+	if err == nil {
+		t.Errorf("Non CM site should not allow Smartstore configuration")
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:356]: ditto:same as above

File: pkg/splunk/enterprise/configuration_test.go [hunk 24]
```diff
@@ -236,3 +236,229 @@ func TestSetVolumeDefault(t *testing.T) {
...
+	}
+}
+
+func TestValidateSplunkSmartstoreSpec(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/enterprise/configuration_test.go:360]: You might want to also test if `smartstore=null`.

File: pkg/apis/enterprise/v1alpha3/common_types.go [hunk 25]
```diff
@@ -67,3 +67,30 @@ type CommonSplunkSpec struct {
...
+	IndexList []IndexSpec `json:"indexes,omitempty"`
+}
+
+// VolumeSpec defines remote volume name and remote volume URI
+type VolumeSpec struct {
+	// Remote volume name
+	Name string `json:"name"`
+
+	// Remote volume URI
+	S3Endpoint string `json:"endpoint"`
```
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/common_types.go:86]: Currently, the config we are supporting is S3 specific, and when we start supporting multiple storages, we may need to name it based on that. One way may be introducing a type of field in addition. We can take a call based on the need at that time?

File: pkg/splunk/enterprise/configuration.go [hunk 26]
```diff
@@ -639,3 +642,69 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
 		podTemplateSpec.Spec.Containers[idx].Env = env
 	}
 }
+
+// LogSmartStoreVolumes logs smartstore volumes
+func LogSmartStoreVolumes(volumeList []enterprisev1.VolumeSpec) {
+	scopedLog := logC.WithName("LogtSmartStoreVolumes")
+	//var temp string
+	for _, volume := range volumeList {
+		scopedLog.Info("Volume: ", "name: ", volume.Name, "endpoint: ", volume.S3Endpoint)
```
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/splunk/enterprise/configuration.go:651]: This is correct according to the log format: "Info(msg string, keysAndValues ...interface{})"

File: pkg/splunk/enterprise/configuration.go [hunk 27]
```diff
@@ -639,3 +642,69 @@ func updateSplunkPodTemplateWithConfig(podTemplateSpec *corev1.PodTemplateSpec,
...
+	}
+}
+
+// LogSmartStoreIndexes logs smartstore indexes
+func LogSmartStoreIndexes(indexList []enterprisev1.IndexSpec) {
+	scopedLog := logC.WithName("LogSmartStoreIndexes")
+	for _, index := range indexList {
+		scopedLog.Info("Index: ", "name: ", index.Name, "location: ", index.RemoteLocation)
```
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/splunk/enterprise/configuration.go:659]: This is correct according to the log format: "Info(msg string, keysAndValues ...interface{})"

File: pkg/splunk/enterprise/configuration.go (hunk 11 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration.go:676]: This is not an error, as not configuring the smart store is perfectly fine(at least for this release)

File: pkg/splunk/enterprise/indexercluster.go (hunk 12 above)
Comment (sgontla) [reply pkg/splunk/enterprise/indexercluster.go:341]: Smartstore is optional for this release.

File: pkg/splunk/enterprise/configuration_test.go (hunk 20 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration_test.go:329]: All these changes are assuming the multi site approach that is what we have today.

File: pkg/splunk/enterprise/configuration_test.go (hunk 21 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration_test.go:332]: same as above: All these changes are assuming the multi site approach that is what we have today.

File: pkg/splunk/enterprise/configuration_test.go (hunk 22 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration_test.go:351]: same as above

File: pkg/splunk/enterprise/configuration.go (hunk 11 above)
Comment (gaurav-splunk) [reply pkg/splunk/enterprise/configuration.go:676]: Ok, makes sense

File: pkg/splunk/enterprise/configuration_test.go (hunk 20 above)
Comment (gaurav-splunk) [reply pkg/splunk/enterprise/configuration_test.go:329]: I think you meant multi-part?

File: pkg/splunk/enterprise/configuration_test.go (hunk 21 above)
Comment (gaurav-splunk) [reply pkg/splunk/enterprise/configuration_test.go:332]: I think you meant multi-part?

File: pkg/apis/enterprise/v1alpha3/common_types.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/apis/enterprise/v1alpha3/common_types.go:86]: Actually, we can remove the S3 like how we removed from Spec, that way we don't need to worry later...

File: pkg/splunk/enterprise/configuration_test.go (hunk 15 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration_test.go:243]: c&p error. Let me fix that.

File: pkg/splunk/enterprise/configuration_test.go (hunk 16 above)
Comment (sgontla) [reply pkg/splunk/enterprise/configuration_test.go:250]: Yes, in this case, its S3 end point. Point is to have some invaid smartstore config. Its tough to get in to all the combinations.
